---
layout: 	post
title: 		"Java编程思想笔记<十二>"
subtitle:	"Error Handling with Exceptions"
date: 		2017-3-12 11:40:00
author: 	"玄天强"
header-img:	"img/mn-bg-xtq.jpg"
catalog: true
tags:
    - Thinking in Java
---
 
*	The ideal time to catch an error is at compile time, before you even try to run the program.
*	抓住错误的理想时机是在编译时，在你试图运行程序前。
*	The rest of the problems must be handled at run time through some formality that allows the originator of the error to pass appropriate information to a recipient who will know how to handle the difficulty properly.
*	剩余问题要在运行期间通过一些形式来处理，该形式允许错误产生者传递合适信息给一个知道要如何处理难题的接受者。
*	To create a robust system, each component must be robust.
*	要创建一个鲁棒的系统，每个部分须是鲁棒的。
*	By providing a consistent error-reporting model using exceptions, Java allows components to reliably communicate problems to client code. 
*	要用异常提供一个一致的错误报告模型，Java允许部件可靠地与客户代码交流问题。

##	Concepts
*	At the point where the problem occurs, you might not know what to do with it, but you do know that you can’t just continue on merrily.
*	在问题发生处，你可能不晓得要做啥，但你知道不能愉快地继续玩耍了。
*	You don’t have enough information in the current context to fix the problem. So you hand the problem out to a higher context where someone is qualified to make the proper decision.
*	在当前上下文你没有足够的信息来解决问题，因此你把它递交给了更高层的上下文，那里有人专门来做合适的决定。
*	You only need to handle the problem in one place, in the so-called *exception handler.*
*	你只需在一个地方处理问题，就是所谓的*exception handler*中。

##	Basic exceptions
*	With an exceptional condition, you cannot continue processing because you don’t have the information necessary to deal with the problem in the current context.
*	在异常条件下，你的程序不能继续，因为当前上下文中你没有处理问题的必要信息。 
*	All you can do is jump out of the current context and relegate that problem to a higher context. This is what happens when you throw an exception.
*	你所要做的只是跳出当前上下文并将问题交给上一层，这是抛出异常时发生的事。
*	You can send information about the error into a larger context by creating an object representing your information and "throwing" it out of your current context. This is called *throwing an exception*.
*	你可以通过创建一个代表你的信息的对象并将其抛出当前上下文来发送错误信息给更大的上下文，这就是*抛出异常*。
*	Exceptions allow you to think of everything that you do as a transaction, and the exceptions guard those transactions.
*	异常使得你将一切视作一个事务，并且异常保护这些事务。
*	Exceptions allow you to (if nothing else) force the program to stop and tell you what went wrong, or (ideally) force the program to deal with the problem and return to a stable state.
*	异常使得你强制程序停止，并告诉你啥玩意有问题了，或者强制程序处理问题并返回到一个稳定状态。

##	Exception arguments
*	There are two constructors in all standard exceptions: The first is the default constructor, and the second takes a string argument so that you can place pertinent information in the exception.
*	标准异常中都有两个构造函数，第一个是默认的，第二个带有一个字符串参数以在异常中放一些相关信息。
*	After creating an exception object with **new**, you give the resulting reference to **throw**.
*	在用new创建了一个异常对象后，你将生成的引用传给throw。
*	A simplistic way to think about exception handling is as a different kind of return mechanism.
*	一种简单的考虑异常处理的方式是将其视为一种不同的返回机制。
*	You can throw any type of Throwable, which is the exception root class.
*	你可以抛出任何Throwble的类型，它是异常的根类。

##	Catching an exception
*	*guarded region* is a section of code that might produce exceptions and is followed by the code to handle those exceptions.
*	*guarded region*是一块可能产生异常的代码，跟随其后的是处理该异常的代码。

###	The try block
*	If you don’t want a **throw** to exit the method, you can set up a special block within that method to capture the exception.
*	如果你不想一个throw退出方法，可以在方法中设置一个特殊块来捕获异常。
*	With exception handling, you put everything in a **try** block and capture all the exceptions in one place.
*	用异常处理，你将所有东西放到一个try块中，并在一个地方捕获所有异常。

###	Exception handlers
*	The thrown exception must end up someplace. This "place" is the *exception handler*.
*	抛出的异常须在某处结束，这个地方就是异常处理器。
*	Sometimes you never use the identifier because the type of the exception gives you enough information to deal with the exception.
*	有时你不用标识符因为异常的类型提供了足够的信息来处理它。
*	If an exception is thrown, the exception-handling mechanism goes hunting for the first handler with an argument that matches the type of the exception. 
*	当一个异常抛出后，异常处理机制查找第一个参数与异常类型匹配的处理器。

###	Termination vs. resumption
*	*termination*: you assume that the error is so critical that there’s no way to get back to where the exception occurred.
*	*termination*:你假定错误太严重了没法返回到异常发生的地方。
*	*resumption*: the exception handler is expected to do something to rectify the situation, and then the faulting method is retried, presuming success the second time.
*	*救活*: 希望异常处理器可以做点啥来拯救，然后重试错误方法，希望第二次可以成功。
*	If you want resumption-like behavior in Java, don’t throw an exception when you encounter an error.
*	要是想在Java中用类似resumption的行为，遇到一个错误时就别抛出异常。
*	Although resumption sounds attractive at first, it isn’t quite so useful in practice.
*	尽管resumption一下子听起来很美好，然而实际上没啥卵用。

##	Creating your own exceptions
*	The most trivial way to create a new type of exception is just to let the compiler create the default constructor for you.
*	最有技巧的创建一个新类型异常的方法是让编译器创建默认构造函数。
*	The most important thing about an exception is the class name.
*	异常最重要的是类名。
*	If you send output to **System.err**, it will not be redirected along with **System.out** so the user is more likely to notice it.
*	如果给System.err发送输出，它不会与System.out一道被重定向，因此使用者更有可能注意到它。  

###	Exceptions and logging
*	It’s more common that you will be catching and logging someone else’s exception, so you must generate the log message in the exception handler.
*	抓住并记录别人家的异常更普遍，因此你须得在异常处理器中生成日志信息。
*	All this dressing-up might be lost on the client programmers using your packages, since they might simply look for the exception to be thrown and nothing more.
*	所有这些打扮在使用你的包的客户程序员那里可能丢失，因为他们可能只是简单地查看要被抛出的异常，其他啥也不做。

##	The exception specification
*	*exception specification*:　Java provides syntax (and forces you to use that syntax) to allow you to politely tell the client programmer what exceptions this method throws.
*	*exception specification*: Java提供语法(并强制使用)来让你心平气和温婉可人地告诉客户程序员这个方法抛出了啥异常。
*	Ｙou must either handle the exception or indicate with an exception specification that it　may be thrown from your method.
*	你必须要么处理异常，要么用异常规范来表明你的方法可能抛出异常。
*	Ｙou can actually start throwing the exception later without requiring changes to　existing code.
*	实际上可以稍后抛出异常而无需改动已有代码。
*	Exceptions that are checked and enforced at compile time are called *checked exceptions*.
*	在编译器检查并强制的异常称为*checked exceptions*。

##	Catching any exception
*	Since the **Exception** class is the base of all the exception classes that are important to the programmer, you don’t get much specific information about the exception, but you can call the methods that come from its base type **Throwable**.
*	由于Exception是所有重要异常类的基类，在你知道的异常信息有限时，你可以调用来着其基类Throwable的方法。  

###	The stack trace
*	**getStackTrace( )** method returns an array of stack trace elements, each representing one stack frame.
*	getStackTrace()方法返回一个stack trace元素的数组，每个代表了一个栈帧。

###	Rethrowing an exception
*	Rethrowing an exception causes it to go to the exception handlers in the nexthigher context.
*	抛出一个异常使得它到了更高层上下文的异常处理器中。
*	Everything about the exception object is preserved, so the handler at the higher context that catches the specific exception type can extract all the information from that object.
*	异常对象的一切都被保留了下来，因此高层上下文中捕获特定异常的处理器可以提取该对象中的所有信息。
*	It’s also possible to rethrow a different exception from the one you caught.
*	也可以从你捕获的异常中再抛出一个不同的异常。 
*	You never have to worry about cleaning up the previous exception, or any exceptions for that matter.
*	永远无需担心对前一个异常或是该事件的任何异常的清理。

###	Exception chaining
*	*exception chainin*: you want to catch one exception and throw another, but still keep the information about the originating exception.
*	*exception chain*: 你打算捕获一个异常并抛出另一个，但依旧保持原始异常的信息。
*	The *cause* is intended to be the originating exception, and by passing it in you maintain the stack trace back to its origin. 
*	*cause*打算作为源异常，通过将它传递进来，你维护了返回到其起始处的stack trace。

##	Standard Java exceptions
*	There are two general types of **Throwable** objects.
*	有两种Throwable对象的一般类型。
*	**Error** represents compile-time and system errors that you don’t worry about catching.
*	Error代表你无需担心捕获的编译期及系统错误。
*	**Exception** is the basic type that can be thrown from any of the standard Java library class methods and from your methods and runtime accidents.
*	Exception是可以从任意标准Java库及自己写的方法中抛出的基本类型。

###	Special case: RuntimeException
*	If any call is made to a null reference, Java will automatically throw a **NullPointerException**.
*	如果调用一个空的引用，Java会自动抛出一个NullPointerExcpetion。
*	You never need to write an exception specification saying that a method might throw a **RuntimeException**, because they are *unchecked exceptions*. 
*	无需写一个异常声明来表明方法可能抛出一个RuntimeException，因为它们是unchecked exceptions。
*	A **RuntimeException**(or anything inherited from it) is a special case, since the compiler doesn’t require an exception specification for these types. 
*	一个RuntimeException(或是任何继承自它的玩意儿)是一个特殊情况，因为对这些类型，编译器不需要异常声明。
*	A **RuntimeException** represents a programming error.
*	一个RuntimeException代表了程序错误。

##	Performing cleanup with finally
*	There’s often some piece of code that you want to execute whether or not an exception is thrown within a **try** block.
*	通常会有些代码不管异常是否在try块中发生你都想执行。
*	Exceptions in Java do not allow you to resume back to where the exception was thrown.
*	Java中的Exceptions不允许你回到异常抛出的地方。

###	What’s finally for?
*	**finally** is important because it allows the programmer to guarantee the release of memory regardless of what happens in the **try** block.
*	finally很重要因为它使得程序员确保内存的释放，不管try块中发生了什么。
*	The **finally** clause is necessary when you need to set something other than memory back to its original state.
*	finally子句在你需要设置一些非内存的东西到其初始状态时很是必须的。
*	Along with the labeled **break** and labeled **continue**, **finally** eliminates the need for a **goto** statement in Java.  
*	与break和continue一道，finally消除了Java中对goto的需要。

###	Using finally during return
*	Because a finally clause is always executed, it’s possible to return from multiple points within a method and still guarantee that important cleanup will be performed.
*	由于finally总是被执行，在方法中的多个点返回并依然保证重要的清理被执行是可能的。

###	Pitfall: the lost exception
*	It’s possible for an exception to simply be lost. This happens with a particular configuration using a **finally** clause.
*	有可能异常会被简单地丢掉，这发生在用一个finally子句做特定配置时。
*	Using ‘return’ inside the **finally** block  will silence any thrown exception.
*	在finally块中用return会使任何异常抛出无效。

##	Exception restrictions
*	When you override a method, you can throw only the exceptions that have been specified in the base-class version of the method. 
*	当你重写一个方法时，你只能抛出基类版本的方法中指定的异常。
*	Interface CANNOT add exceptions to existing methods from the base class, this makes sense because otherwise you’d never know if you were catching the correct thing when working with the base class.
*	接口不能增加异常给基类中的已有方法，这是有意义的，因为否则的话当你在用基类时你永远也不会知道你是否捕捉到了正确的东西。
*	The restriction on exceptions does not apply to constructors. 
*	对异常的限制不会应用到构造函数上。
*	The derived-class constructor must declare any base-class constructor exceptions in its exception specification.
*	继承类的构造函数必须在其异常规范中声明任何基类构造函数的异常。
*	By forcing the derived-class methods to conform to the exception specifications of the base-class methods, substitutability of objects is maintained. 
*	通过强迫继承类的方法来确认基类方法的异常声明，对象的可替性被维护了。
*	A derived-class version of a method may choose not to throw any exceptions, even if the base-class version does.
*	子类方法可以选择不抛出异常，即便父类方法这样做了。
*	The exception specifications are not part of the type of a method, which comprises only the method name and argument types.
*	异常声明不是方法类型的一部分，它只由方法名和参数类型组成。
*	The "exception specification interface" for a particular method may narrow during inheritance and overriding, but it may not widen. 
*	某个特定方法的异常声明接口在继承和重写时可能缩小，但不会扩大。

##	Constructors
*	If you throw an exception from inside a constructor, these cleanup behaviors might not occur properly.
*	如果你在构造函数内抛出一个异常，这些清理行为也许难以正常发生。
*	If a constructor fails partway through its execution, it might not have successfully created some part of the object that will be cleaned up in the **finally** clause.
*	如果构造函数在执行时部分失败了，在finally中将被清理的对象的部分可能还没成功创建。
*	One of the design issues with exceptions is whether to handle an exception completely at this level, to handle it partially and pass the same exception (or a different one) on, or whether to simply pass it on.
*	关于异常的一个设计问题是是否要在本层完全地处理异常，或是部分处理并向上传递相同异常，或是简单地将其向上传递。
*	The basic rule is: Right after you create an object that requires cleanup, begin a **try-finally**.
*	基本规则是:在你创建的需要清理的对象之后开始一个try-finally语句。

##	Exception matching
*	When it finds a match, the exception is considered handled, and no further searching occurs.
*	当发现配准时，考虑对异常进行处理，并且不再进一步搜索。
*	Matching an exception doesn’t require a perfect match between the exception and its handler.
*	匹配一个异常无需异常与其处理器间完美匹配。
*	If you decide to add more derived exceptions to a method, then the client programmer’s code will not need changing as long as the client catches the base-class exceptions.
*	如果你决定给一个方法增加多个继承异常，客户程序员的代码无需改变，只要他捕获了基类的异常。 

##	Alternative approaches
*	It’s worth observing that the issue of programmer convenience in handling errors was a prime motivation for exceptions in the first place.
*	值得注意的是，程序员处理错误时的是否方便是异常的一个主要动机。
*	One of the important goals of exception handling is to move the error-handling code away from the point where the errors occur.
*	异常处理的重要目标之一是将异常处理代码从异常发生处移开。
*	Exception handling also tends to reduce the amount of error-handling code, by allowing one handler to deal with many error sites. 
*	通过允许一个处理器处理多个异常点，异常处理也趋向于减少异常处理代码的数量。
*	Because the compiler forces you to write code right away to handle the exception, this seems like the easiest solution even though it’s probably the worst thing you can do.
*	由于编译器强制你写异常处理的代码，看起来这是最简单的解决方法即便他可能是你能做的最坏的事。

###	History
*	To draw them into a better way of handling errors, the amount of code they would need to "add" must not be onerous.
*	为了给他们一条更好的处理异常的方式，他们要增加的代码量必须不繁杂。
*	The exception specification really has two purposes. 
*	异常声明实际上有两个目的。
*	In C++, no compile-time checking occurs to determine whether or not the function or method will actually throw that exception, or whether the exception specification is complete.
*	C++中没有编译期检测来决定方法或函数是否会真的抛出异常，或是异常声明是否完整。
*	In Java, there are restrictions on the way that Java generics can be used with exception specifications.
*	Java中,泛型与异常声明的方式有限制。

###	Perspectives
*	Java’s important step was to unify the error-reporting model, so that all errors are reported using exceptions.
*	Java中重要的步骤是统一了异常报告模型，由是所有错误都用异常来报告。
*	When Java modified the C++ model so that exceptions were the only way to report errors, the extra enforcement of checked exceptions may have become less necessary.
*	当Java改变了C++模型使得异常是报告错误的唯一方法后，对受检查异常的额外强制就不那么必要了。
*	It’s even more important to realize the limitation of what the compiler is able to do.
*	意识到编译器可以做什么的限制更重要。
*	In any event, the likelihood of checked exceptions ever being removed from Java seems dim.
*	无论如何，去掉Java中受检异常的可能性不大
*	However, if you find that some checked exceptions are getting in your way, or especially if you find yourself being forced to catch exceptions, but you don’t know what to do with them, there are some alternatives.
*	但是，如果你发现一些受检查异常挡了道，或是你发现自己不得不捕获异常，但不晓得该对它们做点啥时，会有其他选择的。

###	Passing exceptions to the console
*	Note that main( ) is also a method that may have an exception specification.
*	注意main()方法也是一个可以有异常声明的方法。
*	By passing it out to the console, you are relieved from writing **try-catch** clauses within the body of **main( )**.
*	通过将它传给控制台，你就从在main()方法中写try-catch块中解放出来了。

###	Converting checked to unchecked exceptions
*	You simply "wrap" a checked exception inside a RuntimeException by passing it to the RuntimeException constructor.
*	通过将其传递给RuntimeExcption的构造函数将一个受检异常包装在一个RuntimeException中。 